samples/python/sqlalchemy-sample/test_add_singer.py

```diff
@@ -0,0 +1,28 @@
+""" Copyright 2022 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from model import Singer
+from sample import add_singer
+from datetime import datetime
+
+singer = Singer()
+singer.id = "123-456-789"
+singer.first_name = "Myfirstname"
+singer.last_name = "Mylastname"
+singer.active = True
+# Manually set a created_at value, as we otherwise do not know which value to
+# add to the mock server.
+singer.created_at = datetime.fromisoformat("2011-11-04T00:05:23.123456+00:00"),
+add_singer(singer)
```

Comment:
Are these test using a framework that I'm unfamiliar with (so something will actually run them as a test and print out results, etc)?

If so, cool!  Do you have a pointer?

If not, I'd encourage making them actual proper runnable tests.  [pytest](https://docs.pytest.org/en/7.2.x/) is probably the most straightforward way to do that with this code.  You'd have to wrap the logic above in a function `def test_<something>():` (for whatever value of `<something>` you prefer.  (Also the file needs to be named as `test_*.py`, but it already is.)  Then you just run `pytest` in the top-level directory, and it should find and run this and all other similarly-constructed tests in this folder, and produce output that's human-readable by default but xUnit-compatible with a flag.  (Note that none of these steps actually require importing pytest in the code in this repository so you don't need to add pytest as a dependency.)

If you want to be able to use pytest but also to be able to run these tests individually without having pytest installed, you could add the following to the bottom of each file:

```
if __name__ == '__main__':
  test_<something>()
```

Or, if you want a little more automation/hackery and you have a bunch of test functions in one file, maybe something like:
```
if __name__ == '__main__':
  for key in globals():
    if key.startswith("test_") and callable(globals()[key]):
      globals()[key]()
```

That way you just have a bunch of files containing functions named `test_*` (which is fine), and if the file is passed directly to the Python interpreter, it will be called.  But pytest will import these files (in which case `__name__` is not `"__main__"`, it's the name of the module), scan them for functions named `test_*`, and run the functions within its framework.

Reply:
Yes, we have a separate test runner for this. We try to use Java for actually running all test cases, so we can then also do all assertions in Java and have one common test framework for all ecosystems. This also has the big advantage that we can use an in-memory mock Spanner server for testing features in all ecosystems.

The sample is for example tested twice:
1. A test runner that uses an in-memory mock Spanner server. We use this server to verify that we see the RPC calls that we actually expect, including the parameters / values that we expect to be passed in. The specific `test_add_singer.py` script is tested [from here](https://github.com/GoogleCloudPlatform/pgadapter/blob/ef00fb99a7a95f0385c0cf5ce788daf12100c2ae/src/test/java/com/google/cloud/spanner/pgadapter/python/sqlalchemy/SqlAlchemySampleTest.java#L517).
2. In addition, we [run the entire sample](https://github.com/GoogleCloudPlatform/pgadapter/blob/ef00fb99a7a95f0385c0cf5ce788daf12100c2ae/src/test/java/com/google/cloud/spanner/pgadapter/python/sqlalchemy/ITSQLAlchemySampleTest.java#L54) against a real Cloud Spanner instance to verify that the queries that are sent are actually supported by Cloud Spanner.